src/rgw/rgw_sal_motr.cc

```diff
@@ -3419,55 +3350,72 @@ int MotrAtomicWriter::complete(size_t accounted_size, const std::string& etag,
-	// TODO: This will be taken care during version implementation.
-    	rc = old_mobj->delete_object(dpp, NULL, y, true);
-	if (rc == 0) {
-      		ldpp_dout(dpp, 20) <<__func__<< ": Old " << obj_type << " ["
-          	<< old_mobj->get_name() <<  "] deleted succesfully" << dendl;
-    	} else {
-      		ldpp_dout(dpp, 0) << __func__<<": Failed to delete old " << obj_type << " ["
-          	<< old_mobj->get_name() <<  "]. Error = " << rc << dendl;
-                // TODO: This will be handled during GC
-    	}
-  	}
-  }
   rc = store->do_idx_op_by_name(bucket_index_iname,
                                 M0_IC_PUT, obj.get_key().to_str(), bl);
-
   if (rc != 0) {
     // TODO: handle this object leak via gc.  
     ldpp_dout(dpp, 0) << __func__ << ": index operation failed, rc = "<< rc << dendl;
     return rc;
   }
-
   store->get_obj_meta_cache()->put(dpp, obj.get_key().to_str(), bl);
 
+  if( !info.versioning_enabled() )
+  {
+    std::string obj_type = "simple object";
+    rgw_bucket_dir_entry old_ent;
+    bufferlist old_check_bl;
+    rc = obj.fetch_null_obj(dpp, old_check_bl);
+    if (rc < 0){
+      ldpp_dout(dpp, 0) <<__func__<< ": Failed to fetch null object, rc : " << rc << dendl;
+      return rc;
+    }
+
+    //TODO: Remove this call in optimization tkt CORTX-31977
+    std::string null_obj_key;
+    rc = obj.fetch_null_obj_reference(dpp, null_obj_key);
+    if (rc < 0){
+      ldpp_dout(dpp, 0) <<__func__<< ": Failed to fetch null reference key, rc : " << rc << dendl;
+      return rc;
+    }
+
+    rc = obj.update_null_reference(dpp, ent);
+    if (rc < 0){
+      ldpp_dout(dpp, 0) <<__func__<< ": Failed to update null reference key, rc : " << rc << dendl;
+      return rc;
+    }
+
+    if (rc == 0 && old_check_bl.length() > 0) {
+      std::unique_ptr<rgw::sal::Object> old_obj = obj.get_bucket()->get_object(rgw_obj_key(obj.get_name()));
+      rgw::sal::MotrObject *old_mobj = static_cast<rgw::sal::MotrObject *>(old_obj.get());
+      auto ent_iter = old_check_bl.cbegin();
+      old_ent.decode(ent_iter);
+      rgw::sal::Attrs attrs;
+      decode(attrs, ent_iter);
+      old_mobj->meta.decode(ent_iter);
+      old_mobj->set_instance(std::move(old_ent.key.instance));
+      if (old_ent.meta.category == RGWObjCategory::MultiMeta)
+          obj_type = "multipart object";
+      ldpp_dout(dpp, 20) <<__func__<< ": Old " << obj_type << " exists" << dendl;
+      // Delete old object
+      // Initialize version of the previous object so that it can delete a specific version,rather than delete the latest object.
+      // TODO: This will be taken care during version implementation.
+      rc = old_mobj->remove_mobj_and_index_entry(dpp, old_ent, null_obj_key);
+      if (rc == 0) {
+          ldpp_dout(dpp, 20) <<__func__<< ": Old " << obj_type << " ["
+            << old_mobj->get_name() <<  "] deleted succesfully" << dendl;
+      } else {
+          ldpp_dout(dpp, 0) << __func__<<": Failed to delete old " << obj_type << " ["
```

Comment:
Error logs after remove_mobj_and_index_entry() all over can be avoided, if we enhance logging within the function.

Retain TODO on line 3414 though